File: packages/typescript/src/client/resources/SetsResource.ts
import type { components } from "../../generated/api";
import type { Page, PaginationOptions } from "../pagination/Page";
import { BaseResource } from "./BaseResource";

// Type aliases
type SetList = components["schemas"]["SetList"];
type SetDetail = components["schemas"]["SetDetail"];
type MasterChecklist = components["schemas"]["MasterChecklist"];
type SetType = components["schemas"]["SetType"];
type CardLocale = components["schemas"]["CardLocale"];

// Response types
type SetListResponse = components["schemas"]["SetListResponse"];
type SetSingleResponse = components["schemas"]["SetSingleResponse"];
type MasterChecklistResponse = components["schemas"]["MasterChecklistResponse"];

/** Options for listing sets */
export interface ListSetsOptions extends PaginationOptions {
  /** Filter by series ID */
  seriesId?: string;
  /** Filter by set type (Base, Expansion, SpecialEdition) */
  setType?: SetType;
  /** Filter by locale (US, UK, Japan, etc.) */
  locale?: CardLocale;
}

/** Options for getting a set by slug */
export interface GetSetBySlugOptions {
  /** Optional series ID to disambiguate slugs */
  seriesId?: string;
  /** Optional locale to disambiguate slugs (default: US) */
  locale?: CardLocale;
}

/** Sets API - browse card sets */
export class SetsResource extends BaseResource {
  /**
   * Get all sets with optional filtering and pagination.
   *
   * @example
   * ```ts
   * // Get first page of sets
   * const page = await client.sets.list();
   *
   * // Filter by series
   * const seriesSets = await client.sets.list({ seriesId: 'series-uuid' });
   *
   * // Iterate through all sets
   * for await (const set of client.sets.list()) {
   *   console.log(set.name);
   * }
   * ```
   */
  async list(options: ListSetsOptions = {}): Promise<Page<SetList>> {
    const query = this.buildQuery({
      page: options.page,
      pageSize: options.pageSize,
      seriesId: options.seriesId,
      setType: options.setType,
      locale: options.locale,
    });

    const response = await this.http.get<SetListResponse>("/Sets", query);

    return this.createPagedResponse(response, (paginationOpts) =>
      this.list({ ...options, ...paginationOpts })
    );
  }

  /**
   * Get a single set by ID.
   *
   * @param id - The set GUID
   */
  async get(id: string): Promise<SetDetail> {
    const response = await this.http.get<SetSingleResponse>(`/Sets/${id}`);

    if (!response.data) {
      throw new Error("Set not found");
    }

    return response.data;
  }

  /**
   * Get a set by its URL-friendly slug.
   *
   * @param slug - The URL-friendly slug (e.g., "151")
   * @param options - Optional series ID or locale to disambiguate
   */
  async getBySlug(
    slug: string,
    options: GetSetBySlugOptions = {}
  ): Promise<SetDetail> {
    const query = this.buildQuery({
      seriesId: options.seriesId,
      locale: options.locale,
    });

    const response = await this.http.get<SetSingleResponse>(
      `/Sets/${slug}`,
      query
    );

    if (!response.data) {
      throw new Error("Set not found");
    }

    return response.data;
  }

  /**
   * Get the master checklist for a set (all cards and variants).
   *
   * @param id - The set GUID
   */
  async getMasterChecklist(id: string): Promise<MasterChecklist> {
    const response = await this.http.get<MasterChecklistResponse>(
      `/Sets/${id}/checklist`
    );

    if (!response.data) {
      throw new Error("Checklist not found");
    }

    return response.data;
  }

  /**
   * Convenience method to iterate through all sets.
   *
   * @example
   * ```ts
   * for await (const set of client.sets.listAll()) {
   *   console.log(set.name);
   * }
   * ```
   */
  async *listAll(
    options: Omit<ListSetsOptions, "page"> = {}
  ): AsyncGenerator<SetList> {
    const page = await this.list({ ...options, page: 1 });
    yield* page;
  }
}

File: packages/csharp/src/PokeForge.SDK/Client/Resources/SetsResource.cs
namespace PokeForge.SDK.Client.Resources;

using System.Runtime.CompilerServices;
using PokeForge.SDK.Client.Http;
using PokeForge.SDK.Client.Pagination;
using PokeForge.SDK.Exceptions;
using PokeForge.SDK.Generated;

/// <summary>
/// Options for listing sets.
/// </summary>
public sealed class ListSetsOptions : PaginationOptions
{
    /// <summary>
    /// Filter by series ID.
    /// </summary>
    public Guid? SeriesId { get; set; }

    /// <summary>
    /// Sort field.
    /// </summary>
    public string? SortBy { get; set; }

    /// <summary>
    /// Sort order (asc or desc).
    /// </summary>
    public string? SortOrder { get; set; }

    /// <summary>
    /// Search by set name.
    /// </summary>
    public string? Search { get; set; }

    internal ListSetsOptions Clone() => new()
    {
        Page = Page,
        PageSize = PageSize,
        SeriesId = SeriesId,
        SortBy = SortBy,
        SortOrder = SortOrder,
        Search = Search
    };
}

/// <summary>
/// Sets API - browse card sets.
/// </summary>
public sealed class SetsResource : BaseResource
{
    internal SetsResource(HttpClientWrapper http) : base(http)
    {
    }

    /// <summary>
    /// Get all sets with optional filtering and pagination.
    /// </summary>
    public async Task<Page<SetList>> ListAsync(
        ListSetsOptions? options = null,
        CancellationToken cancellationToken = default)
    {
        options ??= new ListSetsOptions();

        var query = BuildQuery(
            ("page", options.Page),
            ("pageSize", options.PageSize),
            ("seriesId", options.SeriesId),
            ("sortBy", options.SortBy),
            ("sortOrder", options.SortOrder),
            ("search", options.Search)
        );

        var response = await Http.GetAsync<SetListResponse>("/Sets", query, cancellationToken)
            .ConfigureAwait(false);

        return CreatePagedResponse(
            response!,
            r => r.Data,
            r => r.Pagination,
            async (page, pageSize, ct) =>
            {
                var newOptions = options.Clone();
                newOptions.Page = page;
                newOptions.PageSize = pageSize;
                return await ListAsync(newOptions, ct).ConfigureAwait(false);
            });
    }

    /// <summary>
    /// Get a single set by ID with full details.
    /// </summary>
    public async Task<SetDetail> GetAsync(Guid id, CancellationToken cancellationToken = default)
    {
        var response = await Http.GetAsync<SetSingleResponse>($"/Sets/{id}", cancellationToken: cancellationToken)
            .ConfigureAwait(false);

        return response?.Data ?? throw new NotFoundException("Set not found");
    }

    /// <summary>
    /// Iterate through all sets using IAsyncEnumerable.
    /// </summary>
    public async IAsyncEnumerable<SetList> ListAllAsync(
        ListSetsOptions? options = null,
        [EnumeratorCancellation] CancellationToken cancellationToken = default)
    {
        options ??= new ListSetsOptions();
        options.Page = 1;

        var page = await ListAsync(options, cancellationToken).ConfigureAwait(false);

        await foreach (var set in page.WithCancellation(cancellationToken))
        {
            yield return set;
        }
    }
}
